### PR TITLE
Support various parameter fixing strategies.

### PR DESCRIPTION
**Pull request by mjdenkowski**
This code makes it easier to fix various parameters during training by specifying a named strategy.  Rather than manually specifying each parameter name to freeze, the user can specify a single option that freezes everything except a certain part of the model (`--fixed-param-strategy NAME`).  Most of these simple strategies are taken from "Compact Personalized Models for Neural Machine Translation" ([Wuebker et al., 2018](https://arxiv.org/abs/1811.01990)).  They include:
- Updating only the decoder.
- Updating only the first and last encoder and decoder layer.
- Updating only the embeddings.
- Updating only the output projection layer.

MXNet's convention of specifying which parameters to fix (vs which to train) makes the naming a little awkward, but the options should be clear enough.  This also provides an avenue for adding other parameter fixing strategies in the future.

The log message reporting parameters during training is updated to add the number and percentage of parameters that will be fixed and learned.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] Unit tests covering the new code are added.
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/training.py [hunk 1]
```diff
@@ -67,11 +69,13 @@ def __init__(self,
                  default_bucket_key: Tuple[int, int],
                  bucketing: bool,
                  gradient_compression_params: Optional[Dict[str, Any]] = None,
-                 fixed_param_names: Optional[List[str]] = None) -> None:
+                 fixed_param_names: Optional[List[str]] = None,
+                 fixed_param_strategy: Optional[str] = None) -> None:
```
Comment (fhieber) [inline sockeye/training.py]: can you add a docstring for the new parameter above?

File: sockeye/training.py [hunk 2]
```diff
@@ -183,6 +193,57 @@ def sym_gen(seq_lens):
         self.save_version(self.output_dir)
         self.save_config(self.output_dir)
 
+    def _generate_fixed_param_names(self, param_names: List[str], strategy: str):
```
Comment (fhieber) [inline sockeye/training.py]: Could you add a return type?

File: sockeye/training.py [hunk 3]
```diff
@@ -311,13 +372,23 @@ def log_parameters(self):
...
         for name, array in sorted(arg_params.items()):
             info.append("%s: %s" % (name, array.shape))
-            total_parameters += reduce(lambda x, y: x * y, array.shape)
+            num_parameters = reduce(lambda x, y: x * y, array.shape)
+            total_parameters += num_parameters
+            if name in self.module._fixed_param_names:
+                fixed_parameters += num_parameters
+            else:
+                learned_parameters += num_parameters
+        percent_fixed = 100 * (fixed_parameters / total_parameters)
```
Comment (fhieber) [inline sockeye/training.py]: should we guard against `total_parameters = 0`? Admitted, its very unlikely :)

File: test/unit/test_fixed_param_strategy.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (fhieber) [inline test/unit/test_fixed_param_strategy.py]: 2019

File: sockeye/training.py (hunk 3 above)
Comment (mjdenkowski) [reply sockeye/training.py]: Good catch!

**Pull request merged**